lib/perform_bulk.rb

```diff
@@ -0,0 +1,284 @@
+  #
+  # the above will result in a single job execution, where the *logs job args
+  # are equal to the args of the 25 batched jobs.
+  module Job
+    def self.included(klass)
+      raise ArgumentError, "cannot be used outside of Sidekiq job (got #{klass.ancestors})" unless
+        klass < Sidekiq::Job
+
+      klass.extend ClassMethods
+
+      # NB(ezekg) swap the bulk job's queue over to our capsule's ingress queue in
+      #           case our job doesn't ever call .sidekiq_options
+      klass.sidekiq_options queue: QUEUE_WAITING
+    end
+
+    module ClassMethods
+      # FIXME(ezekg) there's gotta be a better way to do this?
+      def sidekiq_options(...)
+        opts = super(...)
+
+        # hijack bulk job to push new jobs to our capsule's ingress queue, and
+        # then we'll push to the expected egress queue after batching.
+        unless opts['queue'] == QUEUE_WAITING
+          queue_was  = opts.delete('queue') || QUEUE_DEFAULT
+          queue      = QUEUE_WAITING
+
+          Logger.instance.debug(trace: :job) { "hijacking #{name.inspect} from #{queue_was.inspect} to #{queue.inspect}" }
+
+          opts['queue_was'] = queue_was
+          opts['queue']     = queue
+        end
+
+        opts
+      end
+    end
+  end
+
+  class Logger
+    include Singleton
+
+    %i[debug info warn error fatal].each do |level|
+      define_method(level) do |*msgs, **tags, &block|
+        log(level, *msgs, **tags, &block)
+      end
+    end
+
+    private
+
+    def log(level, *msgs, **tags, &block)
+      Sidekiq::Context.with(trace: [LOG_PREFIX, tags.delete(:trace)].compact.join('.'), **tags) do
+        Sidekiq.logger.send(level, *msgs, &block)
+      end
+    end
+  end
+
+  module Logging
+    class Chain
+      def initialize(**tags) = @default_tags = tags.freeze
+
+      %i[debug info warn error fatal].each do |level|
+        define_method(level) do |*msgs, **tags, &block|
+          Logger.instance.send(level, *msgs, **default_tags.merge(tags), &block)
+        end
+      end
+
+      private
+
+      attr_reader :default_tags
+    end
+
+    def self.[](trace, **)
+      Module.new do
+        include Logging
+
+        define_method :logger do
+          @_logger ||= Logging::Chain.new(**, trace:)
+        end
+
+        private :logger
+      end
+    end
+
+    private
+
+    def logger = Logger.instance
+  end
+
+  class Processor
+    include Sidekiq::Job
+    include Logging[:processor]
+
+    sidekiq_options queue: QUEUE_PROCESSING
+
+    def perform(*batch)
+      batch.group_by { _1['class'] }.each do |class_name, job_hashes|
+        args = job_hashes.collect { _1['args'] }
+
+        # NB(ezekg) for a better DX we'll unwrap jobs with a singular arg
+        if args.all?(&:one?)
+          args = args.reduce(&:concat)
+        end
+
+        logger.debug { "batching #{job_hashes.size} #{class_name.inspect} jobs" }
+
+        Runner.perform_async(class_name, args)
+      end
+    end
+  end
+
+  class Runner
```

Comment:
Can we make this so that it shows batch size in Sidekiq logs? e.g. `PerformBulk::Runner[42]: start|done`